OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenHelper.java

```diff
@@ -518,104 +412,128 @@ public String getUserId() throws IOException {
                 break;
             default:
                 throw new IOException("Not supported hash algo!");
         }
 
         // Command APDU for PERFORM SECURITY OPERATION: COMPUTE DIGITAL SIGNATURE (page 37)
-        String apdu =
-                "002A9E9A" // CLA, INS, P1, P2
-                        + dsi // digital signature input
-                        + "00"; // Le
-
-        String response = communicate(apdu);
+        CommandAPDU command = new CommandAPDU(0x00, 0x2A, 0x9E, 0x9A, dsi, MAX_APDU_NE_EXT);
+        ResponseAPDU response = communicate(command);
 
-        if (response.length() < 4) {
-            throw new CardException("Bad response", (short) 0);
+        if (response.getSW() != APDU_SW_SUCCESS) {
+            throw new CardException("Failed to sign", response.getSW());
         }
-        // split up response into signature and status
-        String status = response.substring(response.length() - 4);
-        String signature = response.substring(0, response.length() - 4);
-
-        // while we are getting 0x61 status codes, retrieve more data
-        while (status.substring(0, 2).equals("61")) {
-            Log.d(Constants.TAG, "requesting more data, status " + status);
-            // Send GET RESPONSE command
-            response = communicate("00C00000" + status.substring(2));
-            status = response.substring(response.length() - 4);
-            signature += response.substring(0, response.length() - 4);
-        }
-
-        Log.d(Constants.TAG, "final response:" + status);
 
-        if (!mPw1ValidForMultipleSignatures) {
+        if (!mOpenPgpCapabilities.isPw1ValidForMultipleSignatures()) {
             mPw1ValidatedForSignature = false;
         }
 
-        if (!"9000".equals(status)) {
-            throw new CardException("Bad NFC response code: " + status, parseCardStatus(response));
-        }
+        byte[] signature = response.getData();
 
         // Make sure the signature we received is actually the expected number of bytes long!
-        if (signature.length() != 256 && signature.length() != 512
-                && signature.length() != 768 && signature.length() != 1024) {
-            throw new IOException("Bad signature length! Expected 128/256/384/512 bytes, got " + signature.length() / 2);
+        if (signature.length != 128 && signature.length != 256
+                && signature.length != 384 && signature.length != 512) {
+            throw new IOException("Bad signature length! Expected 128/256/384/512 bytes, got " + signature.length);
         }
 
-        return Hex.decode(signature);
+        return signature;
     }
 
     /**
      * Transceive data via NFC encoded as Hex
      */
-    private String communicate(String apdu) throws IOException {
-        return getHex(mTransport.transceive(Hex.decode(apdu)));
+    private ResponseAPDU communicate(CommandAPDU apdu) throws IOException {
```

Comment:
Considering testability, this method can be neatly split into `sendCommandAndReceiveFirstResponse(CommandAPDU)` and `readFullResponse(ResponseAPDU)`